lib/id3c/cli/command/etl/fhir.py

```diff
@@ -469,6 +472,41 @@ def process_encounter(db: DatabaseSession, encounter: Encounter,
         details         = details)
 
 
+def process_encounter_source(encounter: Encounter) -> Optional[Dict[str, str]]:
+    """
+    Returns the source of the given *encounter*. Splits the source into
+    individual keys from a REDCap URL with the following format:
+
+    https://redcap.iths.org/<project_id>/<record_id>/<event_name>/<repeat_instance>
+
+    Note: <event_name> and <repeat_instance> are only included if the
+    REDCap project has repeating events.
+    """
+    if not encounter.meta:
+        return None
+
+    encounter_source = encounter.meta.source
+
+    if not encounter_source:
+        return None
+
+    # Used named group pattern for easy conversion to dictionary after matching
+    # Include letters a-f in `redcap_record_id` for the barcode record id in UW Retros
+    pattern = re.compile(r'https://redcap\.iths\.org' + \
+                         r'(?P<redcap_project_id>/[0-9]+)' + \
+                         r'(?P<redcap_record_id>/[a-z0-9]+)' + \
+                         r'(?P<redcap_event_name>/\w*)?' + \
+                         r'(?P<redcap_repeat_instance>/[0-9]*)?')
```

Comment:
I just approved the related PR in i3c-customizations, but a thought just occurred to me. 

Instead of constructing a string just to deconstruct it in the next step with regex (which could cause unexpected bugs if the exact formula isn't followed for creating a URI string), why not pass along a dict or something we can parse directly for "project_id", "record_id" etc?

Reply:
The `meta.source` seems to be the only appropriate field to hold provenance data and according to FHIR docs it has to be a URI. 

Would it be better to construct the string with query parameters? So the string would be something like:
```
https://redcap.iths.org/record?project_id=12345&record_id=123&event_name=encounter&repeat_instance=3
```

Reply:
Ah that figures. 

Hmm, clever idea. I _think_ that would end up being more robust. If there was a missing piece, say someone forgot to enter `event_name`, we would have a null value upon ingest rather than every value in the key-value pairs erroneously being shifted up one key.

Reply:
Then again, the regex is probably robust enough that this edge case may not even occur.

Reply:
I agree that it'd be better to use a more standard serialization of our key/value pairs. It will make this ETL code more robust, able to be extended with other keys later, and reduce de-serialization code we have to maintain (e.g. no more regex).

@joverlee521 your suggestion of query params is clever! :-) I wonder about going a step further though and removing the special-casing that even that would require (e.g. around how the URL origin is handled). I think the ideal is that the FHIR ETL routine can deserialize `meta.source` generically and simply pass it into `details`.

My suggestion is a `data:` URI containing JSON, e.g.:

```
data:application/json,{"redcap":{"url":"https://redcap.iths.org","project_id":"...","event_name":"..."}}
```

(The JSON part would be URL-encoded in practice.)

Reply:
Cool! I didn't know you can do this with URIs! I'll update to use this method instead.